php/libraries/CandidateData.class.inc

```diff
@@ -0,0 +1,88 @@
+<?php declare(strict_types=1);
+/**
+ * A CandidateData object is a DTO (data transfer object) to
+ * store a representation of data associated with a Candidate
+ * object in LORIS.
+ *
+ * The CandidateData may be partially loaded, in which case any
+ * getter trying to access an unknown will throw an exception.
+ * This allows modules to only load relevant data in a query,
+ * and avoid the overhead of the Candidate::singleton() constructor.
+ *
+ * For instance, a query that selected CenterID and ProjectID and
+ * timepoint related information:
+ * the database could instantiate a Candidate as:
+ *    $visit = new Candidate(new CandidateData($projectid, $centerid, $timepoints));
+ *
+ * with data loaded from the query, and then call
+ * `$candidate->isAccessibleBy($user)` without having to load all the
+ * other data that Candidate::singleton() does. If an attempt is
+ * made to access a property not in the DTO, an exception is thrown,
+ * rather than silently returning the incorrect value.
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+class CandidateData
+{
+    protected $registrationProjectID;
+    protected $registrationCenterID;
+    protected $timepoints;
+
+    /**
+     * Construct a CandidateData object
+     *
+     * @param ?\ProjectID  $project    The RegistrationProjectID
+     * @param ?\CenterID   $centerID   The RegistrationCenterID
+     * @param ?TimePoint[] $timepoints TimePoint objects for each visit of the
+     *                                 candidate
+     */
+    public function __construct(
+        ?\ProjectID $project,
```

Comment:
why not `$projectID`

Reply:
sure.. I started working on this before named parameters were a thing and just got around to trying to send a PR and wrap it up now.

`$projectID` or `$registrationProjectID`?